ext/drmgr/drmgr.h

```diff
@@ -721,6 +721,93 @@ DR_EXPORT
 ptr_uint_t
 drmgr_reserve_note_range(size_t size);
 
+/***************************************************************************
+ * EMULATION
+ */
+
+/**
+ * Constants used when accessing emulated instruction data with the
+ * `drmgr_get_emulation_instr_data()` function. Each constant refers to an element
+ * of the \p emulated_instr_t struct.
+ */
+typedef enum {
+    DRMGR_EMUL_INSTR_PC,    /**< The PC address of the emulated instruction. */
+    DRMGR_EMUL_INSTR,       /**< The emulated instruction. */
+    DRMGR_EMUL_ISA_VERSION, /**< The version of the ISA the instruction belongs to. */
+} emulated_instr_data_t;
+
+/**
+ * Holds data about an emulated instruction, typically populated by an emulation
+ * client and read by an observational client.
+ */
+typedef struct _emulated_instr_t {
+    app_pc pc;            /**< The PC address of the emulated instruction. */
+    instr_t *instr;       /**< The emulated instruction. */
+    unsigned int version; /**< The version of the ISA the instruction belongs to. */
+    unsigned int fields;  /**< Number of data fields in this struct. */
+} emulated_instr_t;
+
+/**
+ * Inserts a label into \p ilist prior to \p where to indicate the start of a
+ * sequence of instructions emulating an instruction \p instr which does not
+ * exist on the hardware. The label has data attached which describes the
+ * instruction being emulated.
+ *
+ * A label will also appear at the end of the sequence, added using
+ * `drmgr_create_emulation_end()`. These start and stop labels can be detected
+ * by an observational client using `drmgr_is_emulation_start()` and
+ * `drmgr_is_emulation_end()` allowing the client to distinguish between native
+ * app instructions and instructions used for emulation.
+ *
+ * Information about the instruction being emulated can be read from the label using
+ * `drmgr_get_emulation_instr_data()`.
+ *
+ */
+DR_EXPORT
+void
+drmgr_create_emulation_start(void *drcontext, instrlist_t *ilist, instr_t *where,
```

Comment:
It would be good to have a test of these: even if it doesn't do much, just to have a call to this function somewhere will help find simple regressions.

Reply:
I will add a test shortly.